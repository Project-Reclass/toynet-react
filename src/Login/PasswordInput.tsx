/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import React, { FC, useState } from 'react';
import { InputGroup, InputProps, InputRightElement } from '@chakra-ui/react';

import { ToyNetButton, ToyNetInput } from './styled';

const PasswordInput: FC<React.PropsWithChildren<InputProps>> = ({ ...rest }) => {
  const [show, setShow] = useState(false);
  const handleClick = () => setShow(!show);

  return (
    <InputGroup size="md">
      <ToyNetInput
        pr="4.5rem"
        {...rest}
        type={show ? 'text' : 'password'}
        data-testid='password-input'
      />
      <InputRightElement width="4.5rem">
        <ToyNetButton h="1.75rem" size="sm" onClick={handleClick}>
          {show ? 'Hide' : 'Show'}
        </ToyNetButton>
      </InputRightElement>
    </InputGroup>
  );
};

export default PasswordInput;