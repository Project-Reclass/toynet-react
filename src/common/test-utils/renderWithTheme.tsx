import React, { FC } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@chakra-ui/core/dist';
import renderer from 'react-test-renderer';

const Wrapper: FC = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

export const renderWithTheme = (ui: any) => {
  return render(ui, { wrapper: Wrapper });
};

export const renderTreeWithTheme = (ui: any) => {
  return renderer.create(<Wrapper>{ui}</Wrapper>);
};
