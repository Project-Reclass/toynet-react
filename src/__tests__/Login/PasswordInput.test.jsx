import { act, fireEvent } from '@testing-library/react';
import React from 'react';
import { renderWithTheme } from 'src/common/test-utils/renderWithTheme';

import PasswordInput from 'src/Login/PasswordInput';

describe('The password input field', () => {
  it('should match snapshots', () => {
    const { container } =  renderWithTheme(<PasswordInput />);
    expect(container).toMatchSnapshot();
  });
  it('should match snapshot with text hidden by default', () => {
    const { getByTestId, container } =  renderWithTheme(<PasswordInput />);
    const input = getByTestId('password-input');
    act(() => {
      fireEvent.change(input, { target: { value: 'test' } });
    });
    expect(container).toMatchSnapshot();
  });
  it('should match snapshot with text shown', () => {
    const { getByTestId, getByText, container } =  renderWithTheme(<PasswordInput />);

    const btn = getByText(/show/i);
    const input = getByTestId('password-input');

    act(() => {
      fireEvent.change(input, { target: { value: 'test' } });
    });
    act(() => {
      fireEvent.click(btn);
    });

    expect(getByText(/hide/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it('should have a button that toggles from show / hide', () => {
    const { getByText } =  renderWithTheme(<PasswordInput />);

    const btn = getByText(/show/i);
    act(() => {
      fireEvent.click(btn);
    });

    expect(getByText(/hide/i)).toBeInTheDocument();
  });
});