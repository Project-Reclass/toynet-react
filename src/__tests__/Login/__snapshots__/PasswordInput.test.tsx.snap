// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The password input field should match snapshot with text hidden by default 1`] = `
<div>
  <div
    class="css-kjafn5"
  >
    <input
      class="css-1ru8d9z"
      data-testid="password-input"
      type="password"
    />
    <div
      class="css-8ueujo"
    >
      <button
        class="css-13h65yf"
        type="button"
      >
        Show
      </button>
    </div>
  </div>
</div>
`;

exports[`The password input field should match snapshot with text shown 1`] = `
<div>
  <div
    class="css-kjafn5"
  >
    <input
      class="css-1ru8d9z"
      data-testid="password-input"
      type="text"
    />
    <div
      class="css-8ueujo"
    >
      <button
        class="css-13h65yf"
        type="button"
      >
        Hide
      </button>
    </div>
  </div>
</div>
`;

exports[`The password input field should match snapshots 1`] = `
<div>
  <div
    class="css-kjafn5"
  >
    <input
      class="css-1ru8d9z"
      data-testid="password-input"
      type="password"
    />
    <div
      class="css-8ueujo"
    >
      <button
        class="css-13h65yf"
        type="button"
      >
        Show
      </button>
    </div>
  </div>
</div>
`;
